packages/assets-controllers/src/AssetsContractController.ts

```diff
@@ -523,9 +628,12 @@ export class AssetsContractController extends BaseControllerV1<
     tokensToDetect: string[],
     networkClientId?: NetworkClientId,
   ) {
-    const chainId = this.getChainId(networkClientId);
-    const provider = this.getProvider(networkClientId);
-    if (!(chainId in SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID)) {
+    const chainId = this.#getCorrectChainId(networkClientId);
+    const provider = this.#getCorrectProvider(networkClientId);
+    if (
+      !((id): id is keyof typeof SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID =>
+        id in SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID)(chainId)
+    ) {
```

Comment:
Type guard narrows `chainId` to the index signature of `SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID`. Without this, attempting to use `chainId` as a property accessor for `SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID` will result in the following error:

```ts
Element implicitly has an 'any' type because expression of type '`0x${string}`' can't be used to index type '{ readonly "0x1": "0xb1f8e55c7f64d203c1400b9d8555d050f94adf39"; readonly "0x38": "0x2352c63A83f9Fd126af8676146721Fa00924d7e4"; readonly "0x89": "0x2352c63A83f9Fd126af8676146721Fa00924d7e4"; ... 14 more ...; readonly "0x505": "0x6aa75276052d96696134252587894ef5ffa520af"; }'.ts(7053)
```

Reply:
It looks like `hasProperty` only serves as a type guard for `objectToCheck`, and not the property name. The adhoc type guard here narrows `chainId`, not `SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID`. 

`hasProperty(SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID, chainId)` does not `is`-assert that `chainId` is anything narrower than `Hex`.

Reply:
Er, right. That makes sense.

I feel like there must be a more concise way to have TypeScript infer that `id` is `keyof typeof SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID`. But my brain isn't working right now, and this seems readable enough.

Reply:
Yes it seems like just the `in` check should be sufficient to narrow `chainId` down to `keyof`. 

It looks like this update in v5.5 https://github.com/microsoft/TypeScript/pull/57847 might be related.

Reply:
Hmm! Nice, I'll follow that PR.